### PR TITLE
Add box.ctl.promote call to stateful failover

### LINKED ISSUE #1372: Call box.ctl.clear_synchro_queue() on new leader
The external algorithm of  leader selection  needs to take into account the specifics of synchronous replication.

https://github.com/tarantool/doc/issues/1964
https://github.com/tarantool/tarantool/issues/5435

### PR DESCRIPTION
**Pull request by yngvar-antonsson**
I didn't forget about

- [x] Tests
- [x] Changelog
- [x] Documentation

Close #1372

### CODE REVIEW

File: cartridge/failover.lua [hunk 1]
```diff
@@ -420,6 +421,38 @@ local function fencing_start()
...
+
+local function synchro_demote()
+    local box_info = box.info
+    if box_info.synchro ~= nil
+    and box_info.synchro.queue ~= nil
+    and box_info.synchro.queue.owner ~= 0
+    and box_info.synchro.queue.owner == box_info.id then
+        local err = box.ctl.demote()
+        if err ~= nil then
+            return log.error(err)
```
Comment (yngvar-antonsson) [inline cartridge/failover.lua:449]:
```suggestion
            log.error('Failed to demote: %s', err)
```

File: cartridge/failover.lua [hunk 2]
```diff
@@ -420,6 +421,38 @@ local function fencing_start()
     vars.fencing_fiber:name('cartridge.fencing')
 end
 
+local function synchro_promote()
+    if vars.mode == 'stateful'
+    and not vars.single_mode
```
Comment (yngvar-antonsson) [inline cartridge/failover.lua:426]:
```suggestion
```

File: cartridge/failover.lua [hunk 3]
```diff
@@ -420,6 +421,38 @@ local function fencing_start()
...
+    and vars.consistency_needed
+    and vars.cache.is_leader
+    and vars.cache.is_rw
```
Comment (yngvar-antonsson) [inline cartridge/failover.lua:429]:
```suggestion
```

File: cartridge/failover.lua [hunk 4]
```diff
@@ -775,10 +813,13 @@ local function cfg(clusterwide_config, opts)
             -- Replicasets with all_rw flag imply that
             -- consistent switchover isn't necessary
             vars.consistency_needed = false
+            synchro_demote()
         elseif #topology.get_leaders_order(topology_cfg, replicaset_uuid, nil, {only_enabled = true}) == 1 then
             -- Replicaset consists of a single server
             -- consistent switchover isn't necessary
             vars.consistency_needed = false
+            vars.single_mode = true
```
Comment (yngvar-antonsson) [inline cartridge/failover.lua:821]:
```suggestion
```

File: test/integration/failover_synchro_test.lua [hunk 5]
```diff
@@ -0,0 +1,335 @@
...
+
+    kill_server('storage-1')
+
+    local current_master
+    h.retrying({timeout = 20}, function()
+        -- wait until leadeship
+        current_master = get_master(g.cluster, replicaset_uuid)[2]
+        t.assert_not_equals(current_master, storage_1_uuid)
+        g.cluster:server(find_alias_by_uuid(current_master)):exec(function()
+            assert(box.info.ro == false)
```
Comment (yngvar-antonsson) [inline test/integration/failover_synchro_test.lua:197]:
```suggestion
            assert(box.info.ro_reason == nil)
```

File: cartridge/failover.lua [hunk 6]
```diff
@@ -52,6 +52,7 @@ local ValidateConfigError = errors.new_class('ValidateConfigError')
 local StateProviderError = errors.new_class('StateProviderError')
 
 vars:new('mode')
+vars:new('single_mode', false)
```
Comment (yngvar-antonsson) [inline cartridge/failover.lua:55]:
```suggestion
```

**Review by filonenko-mikhail — APPROVED** [review]

**Review by vaintrub — APPROVED** [review]
Thanks for the patch!

**Pull request merged**